File: clipsync-win/ClipSync/Security/TrustStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text.Json;

namespace ClipSync.Security;

public sealed class TrustStore
{
    public sealed record Entry(string DidHex, string Name, DateTime AddedAt);

    private readonly Dictionary<string, Entry> _entries;
    private readonly string _path;
    private readonly object _lock = new();

    private TrustStore(string path, Dictionary<string, Entry> entries)
    {
        _path = path;
        _entries = entries;
    }

    public static TrustStore Load()
    {
        var dir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "ClipSync");
        Directory.CreateDirectory(dir);
        var path = Path.Combine(dir, "trust.json.dpapi");
        if (File.Exists(path))
        {
            try
            {
                var raw = ProtectedData.Unprotect(File.ReadAllBytes(path), null, DataProtectionScope.CurrentUser);
                var map = JsonSerializer.Deserialize<Dictionary<string, Entry>>(raw) ?? new();
                return new TrustStore(path, map);
            }
            catch { /* fall through */ }
        }
        return new TrustStore(path, new Dictionary<string, Entry>());
    }

    public bool IsEmpty { get { lock (_lock) return _entries.Count == 0; } }

    public bool Contains(string didHex)
    {
        lock (_lock) return _entries.ContainsKey(didHex.ToLowerInvariant());
    }

    public bool Contains(byte[] spkiHash)
        => Contains(Convert.ToHexString(spkiHash).ToLowerInvariant());

    public void Add(string didHex, string name)
    {
        lock (_lock)
        {
            _entries[didHex.ToLowerInvariant()] = new Entry(didHex, name, DateTime.UtcNow);
            Persist();
        }
    }

    public void Remove(string didHex)
    {
        lock (_lock)
        {
            _entries.Remove(didHex.ToLowerInvariant());
            Persist();
        }
    }

    /// Forget every trusted peer. After this the device advertises pend=1
    /// again and rejects previously-trusted peers until they re-pair.
    /// Used by the `--reset` command-line switch.
    public void Clear()
    {
        lock (_lock)
        {
            _entries.Clear();
            Persist();
        }
    }

    public IReadOnlyList<Entry> All()
    {
        lock (_lock) return _entries.Values.ToList();
    }

    private void Persist()
    {
        var json = JsonSerializer.SerializeToUtf8Bytes(_entries);
        File.WriteAllBytes(_path, ProtectedData.Protect(json, null, DataProtectionScope.CurrentUser));
    }
}

File: clipsync-mac/Sources/ClipSync/Security/TrustStore.swift
import Foundation
import Security
import Crypto

/// Persistent list of trusted peers, keyed by SPKI SHA-256 hash of the
/// peer's TLS certificate. Stored as a plist at
/// `~/Library/Application Support/ClipSync/trust.plist`.
final class TrustStore {
    struct Entry: Codable {
        var didHex: String          // == hex(spkiHash)
        var name: String
        var addedAt: Date
    }

    private var entries: [String: Entry] = [:]    // keyed by didHex
    private let url: URL
    private let lock = NSLock()

    init(url: URL, entries: [String: Entry]) {
        self.url = url
        self.entries = entries
    }

    static func load() -> TrustStore {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = appSupport.appendingPathComponent("ClipSync", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let url = dir.appendingPathComponent("trust.plist")
        if let data = try? Data(contentsOf: url),
           let entries = try? PropertyListDecoder().decode([String: Entry].self, from: data) {
            return TrustStore(url: url, entries: entries)
        }
        return TrustStore(url: url, entries: [:])
    }

    var isEmpty: Bool { lock.lock(); defer { lock.unlock() }; return entries.isEmpty }

    func contains(hex: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return entries[hex.lowercased()] != nil
    }

    func contains(hash: Data) -> Bool {
        contains(hex: hash.map { String(format: "%02x", $0) }.joined())
    }

    func add(didHex: String, name: String) {
        lock.lock()
        entries[didHex.lowercased()] = Entry(didHex: didHex, name: name, addedAt: Date())
        let snap = entries
        lock.unlock()
        persist(snap)
    }

    func remove(didHex: String) {
        lock.lock()
        entries.removeValue(forKey: didHex.lowercased())
        let snap = entries
        lock.unlock()
        persist(snap)
    }

    /// Forget every trusted peer. After this the device advertises pend=1
    /// again and rejects previously-trusted peers until they re-pair.
    func clear() {
        lock.lock()
        entries.removeAll()
        let snap = entries
        lock.unlock()
        persist(snap)
    }

    /// Clear the persisted trust store on disk, before any instance is
    /// loaded into memory. Used by the `--reset` command-line switch.
    static func reset() {
        load().clear()
        NSLog("--reset cleared trusted peers")
    }

    func all() -> [Entry] {
        lock.lock(); defer { lock.unlock() }
        return Array(entries.values)
    }

    private func persist(_ snap: [String: Entry]) {
        let enc = PropertyListEncoder()
        enc.outputFormat = .binary
        if let data = try? enc.encode(snap) {
            try? data.write(to: url, options: .atomic)
        }
    }

    // MARK: SPKI helper

    /// SHA-256 of the DER SubjectPublicKeyInfo of a certificate.
    static func spkiSha256(of cert: SecCertificate) -> Data {
        // SecCertificateCopyKey extracts the public key; re-encoding as
        // DER SPKI requires the full cert. Easiest: hash the whole cert's
        // public key external representation. This is *not* strictly the
        // SPKI hash but is deterministic for a given key — sufficient for
        // pinning since the cert itself doesn't change.
        guard let key = SecCertificateCopyKey(cert),
              let data = SecKeyCopyExternalRepresentation(key, nil) as Data? else {
            return Data(repeating: 0, count: 32)
        }
        return Data(SHA256.hash(data: data))
    }
}
